docs/CONTRIBUTING.md

```diff
@@ -0,0 +1,513 @@
+
+## Data Input
+
+* **Data from the network, from Bitcoin, and from the config file is untrusted.**  Code that ingests such data _cannot assume anything_ about its structure, and _must_ handle any possible byte sequence that can be submitted to the Stacks node.
+
+* **Data previously written to disk by the node is trusted.** If data loaded from the database that was previously stored by the node is invalid or corrupt, it is appropriate to panic.
+
+* **All input processing is space-bound.**  Every piece of code that ingests data must impose a maximum size on its byte representation.  Any inputs that exceed this size _must be discarded with as little processing as possible_.
+
+* **All input deserialization is resource-bound.** Every piece of code that ingests data must impose a maximum amount of RAM and CPU required to decode it into a structured representation.  If the data does not decode with the allotted resources, then no further processing may be done and the data is discarded.
+
+* **All network input reception is time-bound.**  Every piece of code that ingests data _from the network_ must impose a maximum amount of time that ingestion can take.  If the data takes too long to arrive, then it must be discarded without any further processing.  There is no time bound for data ingested from disk or passed as an argument; this requirement is meant by the space-bound requirement.
+
+* **Untrusted data ingestion must not panic.**  Every piece of code that ingests untrusted data must gracefully handle errors.  Panicking failures are forbidden for such data.  Panics are only allowed if the ingested data was previously written by the node (and thus trusted).
+
+## Non-consensus Changes to Blocks, Microblocks, Transactions, and Clarity
+
+Any changes to code that alters how a block, microblock, or transaction is processed by the node should be **treated as a breaking change until proven otherwise**.  This includes changes to the Clarity VM.  The reviewer _must_ flag any such changes in the PR, and the submitter _must_ convince _all_ reviewers that they will _not_ break consensus.
+
+Changes that touch any of these four code paths must be treated with the utmost care.  If _any_ core developer suspects that a given PR would break consensus, then they _must_ act to prevent the PR from merging.
+
+## Changes to the Peer Network
+
+Any changes to the peer networking code **must be run in production before the PR can be merged.**  The submitter should set up a testable node or set of nodes that reviewers can interact with.
+
+Changes to the peer network should be deployed incrementally and tested by multiple ecosystem entities when possible to verify that they function properly in a production setting.
+
+## Performance Improvements
+
+Any PRs that claim to improve performance **must ship with reproducible benchmarks** that accurately measure the improvement.  This data must also be reported in the PR submission.
```

Comment:
Link to an existing example would help

Reply:
Added in 2b5cda5b6

Reply:
Have a look at how we generate new `costs.clar` contracts.